README.md
# Prometheus Flask Instrumentator

[![PyPI version](https://badge.fury.io/py/prometheus-flask-instrumentator.svg)](https://pypi.python.org/pypi/prometheus-flask-instrumentator/)
[![Maintenance](https://img.shields.io/badge/maintained%3F-yes-green.svg)](https://GitHub.com/Naereen/StrapDown.js/graphs/commit-activity)
[![downloads](https://img.shields.io/pypi/dm/prometheus-flask-instrumentator)](https://pypi.org/project/prometheus-flask-instrumentator/)

![release](https://github.com/trallnag/prometheus-flask-instrumentator/workflows/release/badge.svg)
![test branches](https://github.com/trallnag/prometheus-flask-instrumentator/workflows/test%20branches/badge.svg)
[![codecov](https://codecov.io/gh/trallnag/prometheus-flask-instrumentator/branch/master/graph/badge.svg)](https://codecov.io/gh/trallnag/prometheus-flask-instrumentator)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)

Small package to instrument your Flask app transparently.

    pip install prometheus-flask-instrumentator

## Fast Track

```python
from prometheus_flask_instrumentator import Instrumentator

Instrumentator().instrument(app).expose(app)
```

With this the Flask app is instrumented and all Prometheus metrics can be 
scraped via the `/metrics` endpoint. 

The exporter includes the single metric `http_request_duration_seconds`. 
Basically everything around it can be configured and deactivated. These 
options include:

* Status codes are grouped into `2xx`, `3xx` and so on.
* Requests without a matching template are grouped into the handler `none`.
* Renaming of labels and the metric.
* Regex patterns to ignore certain routes.
* Decimal rounding of latencies.

See the *Example with all parameters* for all possible options or check 
out the documentation itself.

## Example with all parameters

```python
from prometheus_flask_instrumentator import PrometheusFlaskInstrumentator

PrometheusFlaskInstrumentator(
    should_group_status_codes=False,
    should_ignore_untemplated=False,
    should_group_untemplated=False,
    should_round_latency_decimals=True,
    excluded_handlers=[
        "admin",            # Unanchored regex.
        "^/secret/.*$"],    # Full regex example.  
    buckets=(1, 2, 3, 4,),
    metric_name="flask_http"
    label_names=("flask_method", "flask_handler", "flask_status",),
    round_latency_decimals=3,
).instrument(app).expose(app, "/prometheus_metrics")
```

It is important to notice that you don't have to use the `expose()` method if 
adding the endpoint directly to the Flask app does not suit you. There are many 
other ways to expose the metrics.

The defaults are the following:

```python
should_group_status_codes: bool = True,
should_ignore_untemplated: bool = False,
should_group_untemplated: bool = True,
should_round_latency_decimals: bool = False,
excluded_handlers: list = ["/metrics"],
buckets: tuple = Histogram.DEFAULT_BUCKETS,
metric_name: str = "http_request_duration_seconds",
label_names: tuple = ("method", "handler", "status",),
round_latency_decimals: int = 4,
```

## Prerequesites

* `python = "^3.6"` (tested with 3.6 and 3.8)
* `flask = "^1"` (tested with 1.1.2)
* `prometheus-client = "^0.8.0"` (tested with 0.8.0)

## Development

Developing and building this package on a local machine requires 
[Python Poetry](https://python-poetry.org/). I recommend to run Poetry in 
tandem with [Pyenv](https://github.com/pyenv/pyenv). Once the repository is 
cloned, run `poetry install` and `poetry shell`. From here you may start the 
IDE of your choice.

For formatting, the [black formatter](https://github.com/psf/black) is used.
Run `black .` in the repository to reformat source files. It will respect
the black configuration in the `pyproject.toml`.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

* Nothing

## [4.1.1] [4.1.0] 2020-07-15

### Added

* The observed latency values can now be rounded to a certain number of 
    decimals as an opt-in feature. This can improve bytes per sample required in 
    storage solutions like VictoriaMetrics.
* Updates to the documentation.

## [1.0.0] 2020-07-15

### Added

* Method `expose()` implements a way to expose the metrics by adding an 
    endpoint to a given Flask app. Compatible with multiprocess mode.
* Add import alternative `Instrumentator`.

### Changed

* Rename class to `PrometheusFlaskInstrumentator` from `FlaskInstrumentator`.
* Switch to SemVer versioning.
* Split instrumentation and exposition into two parts.
* Moved pass of Flask object from constructor to `instrument()` method.
* Extended testing.

## [20.7.5] [YANKED]

## ... [YANKED]


tests/test_flask_instrumentation.py
import os

import pytest
from flask import Flask
from prometheus_client import REGISTRY

from prometheus_flask_instrumentator import Instrumentator

# ==============================================================================
# Setup

METRIC = "http_request_duration_seconds"
COUNT = f"{METRIC}_counts"
SUM = f"{METRIC}_sum"
BUCKETS = f"{METRIC}_buckets"


def create_app() -> "app":
    app = Flask(__name__)

    # Unregister all collectors.
    collectors = list(REGISTRY._collector_to_names.keys())
    print(f"before unregister collectors={collectors}")
    for collector in collectors:
        REGISTRY.unregister(collector)
    print(f"after unregister collectors={list(REGISTRY._collector_to_names.keys())}")

    # Import default collectors.
    from prometheus_client import (gc_collector, platform_collector,
                                   process_collector)

    # Re-register default collectors.
    process_collector.ProcessCollector()
    platform_collector.PlatformCollector()
    gc_collector.GCCollector()

    @app.route("/")
    def home():
        return "Hello World!"

    @app.route("/path/<page_name>")
    def other_page(page_name):
        return page_name

    @app.route("/to/exclude")
    def exclude():
        return "Exclude me!"

    @app.route("/server_error")
    def server_error():
        raise Exception("Test")
        return "will ever get here"

    @app.route("/ignored")
    @Instrumentator.do_not_track()
    def ignored():
        return "HALLO"

    return app


def get_response(client, path: str) -> "response":
    response = client.get(path)

    print(f"\nResponse  path='{path}' status='{response.status_code}':\n")
    for line in response.data.split(b"\n"):
        print(line.decode())

    return response


def assert_is_not_multiprocess(response) -> None:
    assert response.status_code == 200
    assert b"Multiprocess" not in response.data
    assert b"# HELP process_cpu_seconds_total" in response.data


def assert_request_count(
    expected: float,
    name: str = "http_request_duration_seconds_count",
    handler: str = "/",
    method: str = "GET",
    status: str = "2xx",
) -> None:
    result = REGISTRY.get_sample_value(
        name, {"handler": handler, "method": method, "status": status}
    )
    print(
        (
            f"{name} handler={handler} method={method} status={status} "
            f"result={result} expected={expected}"
        )
    )
    assert result == expected
    assert result + 1.0 != expected


# ==============================================================================
# Tests


def test_app():
    app = create_app()
    client = app.test_client()

    response = get_response(client, "/")
    assert response.status_code == 200
    assert b"Hello World!" in response.data

    response = get_response(client, "/path/whatever")
    assert response.status_code == 200
    assert b"whatever" in response.data

    response = get_response(client, "/metrics")
    assert response.status_code == 404


def test_metrics_endpoint_availability():
    app = create_app()
    Instrumentator().instrument(app).expose(app)
    client = app.test_client()

    get_response(client, "/")

    response = get_response(client, "/metrics")
    assert_is_not_multiprocess(response)
    assert_request_count(1)
    assert METRIC.encode() in response.data


def test_parameter_existence():
    app = create_app()
    instrumentator = Instrumentator().instrument(app)

    assert hasattr(instrumentator, "should_group_status_codes")
    assert instrumentator.should_group_status_codes is True

    assert hasattr(instrumentator, "should_ignore_untemplated")
    assert instrumentator.should_ignore_untemplated is False

    assert hasattr(instrumentator, "should_group_untemplated")
    assert instrumentator.should_group_untemplated is True


def test_label_names():
    app = create_app()
    Instrumentator().instrument(app).expose(app)
    client = app.test_client()

    get_response(client, "/")

    response = get_response(client, "/metrics")
    assert_is_not_multiprocess(response)
    assert_request_count(1)
    assert b'method="' in response.data
    assert b'handler="' in response.data
    assert b'status="' in response.data


# ------------------------------------------------------------------------------
# Test status code. should_group_status_codes.


def test_grouped_status_codes():
    app = create_app()
    Instrumentator().instrument(app).expose(app)
    client = app.test_client()

    client.post("/")  # -> 405 -> 4xx
    client.post("/")  # -> 405 -> 4xx
    client.get("/")  # -> 200 -> 2xx
    client.get("/")  # -> 200 -> 2xx
    client.get("/")  # -> 200 -> 2xx

    response = get_response(client, "/metrics")
    assert_is_not_multiprocess(response)
    assert_request_count(3)
    assert b'status="2xx"' in response.data
    assert b'status="4xx"' in response.data
    assert b'status="405"' not in response.data


def test_ungrouped_status_codes():
    app = create_app()
    Instrumentator(should_group_status_codes=False).instrument(app).expose(app)
    client = app.test_client()

    client.post("/")  # -> 405
    client.post("/")  # -> 405
    client.get("/")  # -> 200

    response = get_response(client, "/metrics")
    assert b'status="2xx"' not in response.data
    assert b'status="200"' in response.data
    assert b'status="4xx"' not in response.data
    assert b'status="405"' in response.data


# ------------------------------------------------------------------------------
# Test handler templation handling.


def test_ignore_untemplated():
    app = create_app()
    Instrumentator(should_ignore_untemplated=True).instrument(app).expose(app)
    client = app.test_client()

    get_response(client, "/")  # Exists
    get_response(client, "/fefwefwe4533")  # Does not exist
    get_response(client, "/booogo")  # Does not exist

    response = get_response(client, "/metrics")
    assert b'status="4xx"' not in response.data
    assert b'status="404"' not in response.data
    assert b'status="2xx"' in response.data
    assert b'handler="/fefwefwe4533"' not in response.data
    assert b'handler="none"' not in response.data


def test_include_untemplated_dont_group():
    app = create_app()
    Instrumentator(should_group_untemplated=False).instrument(app).expose(app)
    client = app.test_client()

    get_response(client, "/")  # Exists
    get_response(client, "/this_does_not_exist")

    response = get_response(client, "/metrics")
    assert b'handler="/this_does_not_exist"' in response.data
    assert b'handler="none"' not in response.data
    assert b'status="4xx"' in response.data


def test_include_untemplated_group():
    app = create_app()
    Instrumentator().instrument(app).expose(app)
    client = app.test_client()

    get_response(client, "/")  # Exists
    get_response(client, "/this_does_not_exist")

    response = get_response(client, "/metrics")
    assert b'handler="/this_does_not_exist"' not in response.data
    assert b'handler="none"' in response.data
    assert b'status="4xx"' in response.data


# ------------------------------------------------------------------------------
# Test label names


def test_default_label_names():
    app = create_app()
    instrumentator = Instrumentator()
    instrumentator.instrument(app).expose(app)
    client = app.test_client()

    client.get("/")

    response = get_response(client, "/metrics")

    for label in instrumentator.label_names:
        assert f'{label}="'.encode() in response.data

    for label in ("endpoint", "path", "status_code"):
        assert f'{label}="'.encode() not in response.data


def test_custom_label_names():
    app = create_app()
    instrumentator = Instrumentator(label_names=("a", "b", "c",))
    instrumentator.instrument(app).expose(app)
    client = app.test_client()

    client.get("/")

    response = get_response(client, "/metrics")

    for label in (
        "a",
        "b",
        "c",
    ):
        assert f'{label}="'.encode() in response.data

    for label in (
        "method",
        "handler",
        "status",
    ):
        assert f'{label}="'.encode() not in response.data


# ------------------------------------------------------------------------------
# Test exclusion of paths


def test_do_not_track_decorator():
    app = create_app()
    instrumentator = Instrumentator(excluded_handlers=[])
    instrumentator.instrument(app).expose(app)
    client = app.test_client()

    client.get("/ignored")
    client.get("/ignored")
    client.get("/")

    response = get_response(client, "/metrics")
    assert b'handler="/ignored"' not in response.data
    assert b'handler="/"' in response.data


def test_exclude_paths():
    app = create_app()
    instrumentator = Instrumentator(excluded_handlers=["/to/exclude"])
    instrumentator.instrument(app).expose(app)
    client = app.test_client()

    client.get("/")
    client.get("/to/exclude")
    client.get("/to/exclude")

    response = get_response(client, "/metrics")
    assert b'handler="/"' in response.data
    assert b'handler="/to/exclude"' not in response.data


# ------------------------------------------------------------------------------


def test_bucket_without_inf():
    app = create_app()
    Instrumentator(buckets=(1, 2, 3,)).instrument(app).expose(app)
    client = app.test_client()

    client.get("/")

    response = get_response(client, "/metrics")
    assert b"http_request_duration_seconds_bucket" in response.data


# ------------------------------------------------------------------------------


def test_unhandled_server_error():
    app = create_app()
    Instrumentator().instrument(app).expose(app)
    client = app.test_client()

    client.get("/")

    response = get_response(client, "/server_error")
    assert response.status_code == 500

    response = get_response(client, "/metrics")
    assert b'handler="/server_error"' in response.data
    assert b'status="5xx"' in response.data


# ------------------------------------------------------------------------------


def test_custom_endpoint():
    app = create_app()
    Instrumentator().instrument(app).expose(app, "/custom_metrics")
    client = app.test_client()

    client.get("/")

    response = get_response(client, "/metrics")
    assert b"http_request_duration_seconds_bucket" not in response.data

    response = get_response(client, "/custom_metrics")
    assert b"http_request_duration_seconds_bucket" in response.data


# ------------------------------------------------------------------------------


def test_custom_metric_name():
    app = create_app()
    Instrumentator(metric_name="xzy").instrument(app).expose(app)
    client = app.test_client()

    client.get("/")

    response = get_response(client, "/metrics")
    assert b"http_request_duration_seconds_bucket" not in response.data
    assert b"xzy_bucket" in response.data


# ------------------------------------------------------------------------------
# Test decimal rounding.


def calc_entropy(decimal_str: str):
    decimals = [int(x) for x in decimal_str]
    print(decimals)
    entropy = 0
    for i in range(len(decimals)):
        if i != 0:
            entropy += abs(decimals[i] - decimals[i - 1])
    return entropy


def test_entropy():
    assert calc_entropy([1, 0, 0, 4, 0]) == 9


def test_default_no_rounding():
    app = create_app()
    Instrumentator().instrument(app).expose(app)
    client = app.test_client()

    client.get("/")
    client.get("/")
    client.get("/")

    _ = get_response(client, "/metrics")

    result = REGISTRY.get_sample_value(
        "http_request_duration_seconds_sum",
        {"handler": "/", "method": "GET", "status": "2xx"},
    )

    entropy = calc_entropy(str(result).split(".")[1][4:])

    assert entropy > 15


def test_rounding():
    app = create_app()
    Instrumentator(should_round_latency_decimals=True).instrument(app).expose(app)
    client = app.test_client()

    client.get("/")
    client.get("/")
    client.get("/")

    _ = get_response(client, "/metrics")

    result = REGISTRY.get_sample_value(
        "http_request_duration_seconds_sum",
        {"handler": "/", "method": "GET", "status": "2xx"},
    )

    entropy = calc_entropy(str(result).split(".")[1][4:])

    assert entropy < 10


# ------------------------------------------------------------------------------


def is_prometheus_multiproc_set():
    if "prometheus_multiproc_dir" in os.environ:
        pmd = os.environ["prometheus_multiproc_dir"]
        if os.path.isdir(pmd):
            return True
    else:
        return False


# The environment variable MUST be set before anything regarding Prometheus is
# imported. That is why we cannot simply use `tempfile` or the fixtures
# provided by pytest. Test with:
#       mkdir -p /tmp/test_multiproc;
#       export prometheus_multiproc_dir=/tmp/test_multiproc;
#       pytest -k test_multiprocess_with_var_set;
#       rm -rf /tmp/test_multiproc;
#       unset prometheus_multiproc_dir


@pytest.mark.skipif(
    is_prometheus_multiproc_set() is False,
    reason="Environment variable must be set before starting Python process.",
)
def test_multiprocess_with_var_set():
    app = create_app()
    Instrumentator().instrument(app).expose(app)
    client = app.test_client()

    get_response(client, "/")

    response = get_response(client, "/metrics")
    assert response.status_code == 200
    assert b"Multiprocess" in response.data
    assert b"# HELP process_cpu_seconds_total" not in response.data
    assert b"http_request_duration_seconds" in response.data


@pytest.mark.skipif(
    is_prometheus_multiproc_set() is True, reason="Just test handling of env detection."
)
def test_multiprocess_with_var_not_set(monkeypatch, tmp_path):
    monkeypatch.setenv("prometheus_multiproc_dir", "DOES/NOT/EXIST")

    app = create_app()
    with pytest.raises(Exception):
        Instrumentator(buckets=(1, 2, 3,)).instrument(app).expose(app)


prometheus_flask_instrumentator/flask_instrumentation.py
import re
import os
from typing import Tuple
from functools import wraps
from timeit import default_timer

from prometheus_client import Histogram
from flask import Flask, request


class FlaskInstrumentator:
    def __init__(
        self,
        should_group_status_codes: bool = True,
        should_ignore_untemplated: bool = False,
        should_group_untemplated: bool = True,
        excluded_handlers: list = ["/metrics"],
        buckets: tuple = Histogram.DEFAULT_BUCKETS,
        label_names: tuple = ("method", "handler", "status",),
    ):
        """
        :param should_group_status_codes: Groups all status codes into `1xx`, `2xx` 
            and so on.

        :param should_ignore_untemplated: Should a request to a non-existing handler
            be ignored or not? By default False.

        :param should_group_untemplated: Should requests without a matching 
            template be grouped to handler None? Defaults to True.

        :param excluded_handlers: This list of strings will be regex. compiled. 
            Matched patterns will not be recorded. Defaults to ["/metrics"].

        :param buckets: Override default buckets. Defaults to Prometheus 
            histogram default.
        
        :param label_names: Sets the labelnames of the metric. `x[0]` -> `POST`, 
            `PUT` etc. `x[1]` -> `/getorder`, `/login` etc. `x[2]` -> `500`.         
        """

        self.should_group_status_codes = should_group_status_codes
        self.should_ignore_untemplated = should_ignore_untemplated
        self.should_group_untemplated = should_group_untemplated

        if excluded_handlers:
            self.excluded_handlers = [re.compile(path) for path in excluded_handlers]
        else:
            self.excluded_handlers = []

        if buckets[-1] == float("inf"):
            self.buckets = buckets
        else:
            self.buckets = buckets + (float("inf"),)

        self.label_names = label_names

    def instrument(self, app: Flask) -> "self":
        """Performs the actual instrumentation by using Flask hooks.
        
        :param app: Flask application to be instrumented.
        :return: self.
        """

        histogram = Histogram(
            name="http_request_duration_seconds",
            documentation="Duration of HTTP requests in seconds",
            labelnames=self.label_names,
            buckets=self.buckets,
        )

        @app.before_request
        def act_before_request():
            if self._shall_be_ignored(request):
                return

            request._custom_start_time = default_timer()

        @app.after_request
        def act_after_request(response):
            if self._shall_be_ignored(request):
                return response

            total_time = max(default_timer() - request._custom_start_time, 0)

            histogram.labels(
                *self._create_label_tuple(
                    request.method,
                    request.url_rule,
                    request.path,
                    str(response.status_code),
                )
            ).observe(total_time)

            return response

        @app.teardown_request
        def act_on_teardown_request(exception=None):
            if not exception or self._shall_be_ignored(request):
                return

            total_time = max(default_timer() - request._custom_start_time, 0)

            histogram.labels(
                *self._create_label_tuple(
                    request.method, request.url_rule, request.path, "500"
                )
            ).observe(total_time)

        return self

    def expose(self, app: Flask, endpoint: str = "/metrics") -> "self":
        """Exposes Prometheus metrics by adding endpoint to the given app.

        **Important**: There are many different ways to expose metrics. This is 
        just one of them, suited for both multiprocess and singleprocess mode. 
        Refer to the Prometheus Python client documentation for more information.

        :param app: Flask app where the endpoint should be added to.
        :param endpoint: Route of the endpoint. Defaults to "/metrics".
        :param return: self.
        """

        from prometheus_client import REGISTRY, CONTENT_TYPE_LATEST, generate_latest
        from prometheus_client import multiprocess, CollectorRegistry

        if "prometheus_multiproc_dir" in os.environ:
            pmd = os.environ["prometheus_multiproc_dir"]
            if os.path.isdir(pmd):
                registry = CollectorRegistry()
                multiprocess.MultiProcessCollector(registry)
            else:
                raise ValueError(
                    f"Env var prometheus_multiproc_dir='{pmd}' not a directory."
                )
        else:
            registry = REGISTRY

        @app.route(endpoint)
        def metrics():
            data = generate_latest(REGISTRY)
            headers = {
                "Content-Type": CONTENT_TYPE_LATEST,
                "Content-Length": str(len(data)),
            }
            return data, 200, headers

    def _create_label_tuple(
        self, method: str, url_rule: str, url_path: str, code: str
    ) -> Tuple[str, str, str]:
        """Processes label values based on config."""

        if self.should_group_status_codes:
            code = code[0] + "xx"

        # 'self.should_ignore_untemplated' will always be 'False'

        if url_rule:
            handler = url_rule
        elif self.should_group_untemplated:
            handler = "none"
        else:
            handler = url_path

        return (
            method,
            handler,
            code,
        )

    def _shall_be_ignored(self, request) -> bool:
        """Decides if the request should be ignored or not.
        
        It first checks for the `_pfi_ignore` attribute to reduce CPU cycles in 
        subsequent runs.
        """

        if hasattr(request, "_pfi_ignore") and request._pfi_ignore:
            return True

        if any(p.search(request.path) for p in self.excluded_handlers):
            request._pfi_ignore = True
            return True

        if self.should_ignore_untemplated and not request.url_rule:
            request._pfi_ignore = True
            return True

        return False

    @staticmethod
    def do_not_track():
        def decorator(f):
            @wraps(f)
            def wrapper(*args, **kwargs):
                request._pfi_ignore = True
                return f(*args, **kwargs)

            return wrapper

        return decorator


prometheus_flask_instrumentator/instrumentation.py
import os
import re
from functools import wraps
from timeit import default_timer
from typing import Tuple

from flask import Flask, request
from prometheus_client import Histogram


class PrometheusFlaskInstrumentator:
    def __init__(
        self,
        should_group_status_codes: bool = True,
        should_ignore_untemplated: bool = False,
        should_group_untemplated: bool = True,
        should_round_latency_decimals: bool = False,
        excluded_handlers: list = ["/metrics"],
        buckets: tuple = Histogram.DEFAULT_BUCKETS,
        metric_name: str = "http_request_duration_seconds",
        label_names: tuple = ("method", "handler", "status",),
        round_latency_decimals: int = 4,
    ):
        """
        :param should_group_status_codes: Groups all status codes into `1xx`, `2xx` 
            and so on.

        :param should_ignore_untemplated: Should a request to a non-existing handler
            be ignored or not? By default False.

        :param should_group_untemplated: Should requests without a matching 
            template be grouped to handler None? Defaults to True.

        :param should_round_latency_decimals: Should recorded latencies be 
            rounded to a certain number of decimals?

        :param excluded_handlers: This list of strings will be regex. compiled. 
            Matched patterns will not be recorded. Defaults to ["/metrics"].

        :param buckets: Override default buckets. Defaults to Prometheus 
            histogram default.

        :param metric_name: Name of the latency metric. Defaults to 
            "http_request_duration_seconds".
        
        :param label_names: Sets the labelnames of the metric. `x[0]` -> `POST`, 
            `PUT` etc. `x[1]` -> `/getorder`, `/login` etc. `x[2]` -> `500`.     

        :param round_latency_decimals: Number of decimals latencies should be 
            rounded to, provided `should_round_latency_decimals` is `True    
        """

        self.should_group_status_codes = should_group_status_codes
        self.should_ignore_untemplated = should_ignore_untemplated
        self.should_group_untemplated = should_group_untemplated
        self.should_round_latency_decimals = should_round_latency_decimals

        if excluded_handlers:
            self.excluded_handlers = [re.compile(path) for path in excluded_handlers]
        else:
            self.excluded_handlers = []

        if buckets[-1] == float("inf"):
            self.buckets = buckets
        else:
            self.buckets = buckets + (float("inf"),)

        self.metric_name = metric_name
        self.label_names = label_names
        self.round_latency_decimals = round_latency_decimals

    def instrument(self, app: Flask) -> "self":
        """Performs the actual instrumentation by using Flask hooks.
        
        :param app: Flask application to be instrumented.
        :return: self.
        """

        histogram = Histogram(
            name=self.metric_name,
            documentation="Duration of HTTP requests in seconds",
            labelnames=self.label_names,
            buckets=self.buckets,
        )

        @app.before_request
        def act_before_request():
            if self._shall_be_ignored(request):
                return

            request._custom_start_time = default_timer()

        @app.after_request
        def act_after_request(response):
            if self._shall_be_ignored(request):
                return response

            total_time = max(default_timer() - request._custom_start_time, 0)

            if self.should_round_latency_decimals:
                total_time = round(total_time, self.round_latency_decimals)

            histogram.labels(
                *self._create_label_tuple(
                    request.method,
                    request.url_rule,
                    request.path,
                    str(response.status_code),
                )
            ).observe(total_time)

            return response

        @app.teardown_request
        def act_on_teardown_request(exception=None):
            if not exception or self._shall_be_ignored(request):
                return

            total_time = max(default_timer() - request._custom_start_time, 0)

            if self.should_round_latency_decimals:
                total_time = round(total_time, self.round_latency_decimals)

            histogram.labels(
                *self._create_label_tuple(
                    request.method, request.url_rule, request.path, "500"
                )
            ).observe(total_time)

        return self

    def expose(self, app: Flask, endpoint: str = "/metrics") -> "self":
        """Exposes Prometheus metrics by adding endpoint to the given app.

        **Important**: There are many different ways to expose metrics. This is 
        just one of them, suited for both multiprocess and singleprocess mode. 
        Refer to the Prometheus Python client documentation for more information.

        :param app: Flask app where the endpoint should be added to.
        :param endpoint: Route of the endpoint. Defaults to "/metrics".
        :param return: self.
        """

        from prometheus_client import (CONTENT_TYPE_LATEST, REGISTRY,
                                       CollectorRegistry, generate_latest,
                                       multiprocess)

        if "prometheus_multiproc_dir" in os.environ:
            pmd = os.environ["prometheus_multiproc_dir"]
            if os.path.isdir(pmd):
                registry = CollectorRegistry()
                multiprocess.MultiProcessCollector(registry)
            else:
                raise ValueError(
                    f"Env var prometheus_multiproc_dir='{pmd}' not a directory."
                )
        else:
            registry = REGISTRY

        @app.route(endpoint)
        def metrics():
            data = generate_latest(registry)
            headers = {
                "Content-Type": CONTENT_TYPE_LATEST,
                "Content-Length": str(len(data)),
            }
            return data, 200, headers

    def _create_label_tuple(
        self, method: str, url_rule: str, url_path: str, code: str
    ) -> Tuple[str, str, str]:
        """Processes label values based on config."""

        if self.should_group_status_codes:
            code = code[0] + "xx"

        # 'self.should_ignore_untemplated' will always be 'False'

        if url_rule:
            handler = url_rule
        elif self.should_group_untemplated:
            handler = "none"
        else:
            handler = url_path

        return (
            method,
            handler,
            code,
        )

    def _shall_be_ignored(self, request) -> bool:
        """Decides if the request should be ignored or not.
        
        It first checks for the `_pfi_ignore` attribute to reduce CPU cycles in 
        subsequent runs.
        """

        if hasattr(request, "_pfi_ignore") and request._pfi_ignore:
            return True

        if any(p.search(request.path) for p in self.excluded_handlers):
            request._pfi_ignore = True
            return True

        if self.should_ignore_untemplated and not request.url_rule:
            request._pfi_ignore = True
            return True

        return False

    @staticmethod
    def do_not_track():
        def decorator(f):
            @wraps(f)
            def wrapper(*args, **kwargs):
                request._pfi_ignore = True
                return f(*args, **kwargs)

            return wrapper

        return decorator
